Add swarm mode support

ketan opened the pull request:
This was based off #491, with a few changes -
* all commits were squashed into one
* the `SwarmModeDockerClient` and `DefaultSwarmModeDockerClient` subclasses
  were removed and pushed into `DockerClient` and `DefaultDockerClient`

ketan commented:
I'm at a loss to understand why checkstyle is failing even on master. But the build is apparently green. https://travis-ci.org/spotify/docker-client/branches

Does someone have any pointers?

johnflavin commented:
I looked at your latest builds. Checkstyle is passing, so I guess you managed to fix whatever it found distasteful.

But all the tests are failing. Like, every one. I think it could be the unix socket connection. All the test failures are trying to connect to `unix://localhost:80`. Typically the socket connections are to `unix:///var/run/docker.sock`. 

(Total speculation follows:) are you testing this code locally with your docker daemon open to http connections? Because if I remember correctly the daemon on the test machines is not; the only way to connect is `/var/run/docker.sock`.

ketan commented:
Pretty sure I did not change any setup, in fact I'm on a Mac, so nothing
running on port 80. I suspect it is Travis doing something funny.

Will investigate further.

On Thu, Sep 22, 2016, 4:37 PM Flavin <notifications@github.com> wrote:

> I looked at your latest builds. Checkstyle is passing, so I guess you
> managed to fix whatever it found distasteful.
>
> But all the tests are failing. Like, every one. I think it could be the
> unix socket connection. All the test failures are trying to connect to
> unix://localhost:80. Typically the socket connections are to
> unix:///var/run/docker.sock.
>
> (Total speculation follows:) are you testing this code locally with your
> docker daemon open to http connections? Because if I remember correctly the
> daemon on the test machines is not; the only way to connect is
> /var/run/docker.sock.
>
> —
> You are receiving this because you authored the thread.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/spotify/docker-client/pull/499#issuecomment-248872080>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAApZmgi0HiNKRyl4S9JxPWI2vqUw9Wlks5qsmFmgaJpZM4KDrKK>
> .
>

ketan commented:
Allright, the build is now green, however the coverage is down because of the plethora of classes that represent the JSON data.

If all is good, can I expect a 5.0.3 release?

ketan commented:
BTW — support for `docker node` and friends is not part of this PR, if you're OK merging this PR, I'll submit another PR for `docker node` support.

ketan commented on src/main/java/com/spotify/docker/client/messages/mount/Driver.java:
```diff
@@ -0,0 +1,97 @@
+/*
```
Looking at the api docs, for a few references to where this class is used, it appears that the class is just a generic thing that describes a driver name, and options for a variety of things, including networks and tasks.

ketan commented:
@davidxia — I've changed all timestamp fields from `String` to `java.util.Date`.

> In src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java:
> 
> > @@ -3041,6 +3064,198 @@ public void testIpcMode() throws Exception {
>  
>    }
>  
> +  @Test
> +  public void testInspectSwarm() throws Exception {
> +    requireDockerApiVersionAtLeast("1.24", "swarm support");
> +
> +    final Swarm swarm = sut.inspectSwarm();
> +    assertThat(swarm.createdAt(), is(not(isEmptyOrNullString())));
> Should we check not null for the other fields? Seems like we should check for ID at least.
> 
> 

I'm not sure what "value" asserting on fields adds, esp for code that appears to declarative in nature. I've added a test to assert the swarm `id` attribute as you requested.

ketan commented:
Also, I could not help but notice that the indentation in most of the code is off. Some files are indented using 2 spaces, others with 4 spaces. What's the deal?

ketan commented:
All tests are green now

mattnworb commented:
Since a chunk of the code in this PR was originally contributed by @derekadams, is there any way to maintain that info in the commit author field?

ketan commented:
I'm not sure how to go about putting multiple authors in the author field. I'm happy to attribute the entire commit to him, I don't care so much about my attribution.

Let me know what you'd prefer, I'd really like to get this merged ASAP :)

johnflavin commented:
This looks like it will do the trick: http://stackoverflow.com/a/31669789/1288474

ketan commented:
@johnflavin — Done, travis is green.

ketan commented:
Bump!

davidxia commented on src/main/java/com/spotify/docker/client/DefaultDockerClient.java:
```diff
@@ -1539,6 +1552,163 @@ public LogStream execStart(final String execId, final ExecStartParameter... para
+          throws DockerException, InterruptedException {
+    assertAPIVersionIsAbove("1.24");
+    WebTarget resource = resource().path("services");
+    final Map<String, List<String>> filters = new HashMap<String, List<String>>();
```
You can use the diamond type `<>` here. The pom.xml specifies Java 1.7.

davidxia commented on src/main/java/com/spotify/docker/client/DefaultDockerClient.java:
```diff
@@ -1539,6 +1552,163 @@ public LogStream execStart(final String execId, final ExecStartParameter... para
+  }
+
+  @Override
+  public ServiceCreateResponse createService(ServiceSpec spec, ServiceCreateOptions options)
```
nitpick: Can you use `final` for all the parameter arguments in this class?

davidxia commented on src/main/java/com/spotify/docker/client/DefaultDockerClient.java:
```diff
@@ -1539,6 +1552,163 @@ public LogStream execStart(final String execId, final ExecStartParameter... para
+  }
+
+  @Override
+  public void updateService(String serviceId, Long version, ServiceSpec spec)
```
Same as comment here https://github.com/spotify/docker-client/pull/499/files#diff-163ee817c343ffd4cbc08d4c3412a8a4R1563

davidxia commented on src/main/java/com/spotify/docker/client/messages/swarm/Service.java:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    public static class CriteriaBuilder {
```
For consistency, can you use the builder pattern in `ContainerConfig`? Also make sure the access modifiers for classes and methods are as strict as they can be.

davidxia commented on src/main/java/com/spotify/docker/client/messages/swarm/Task.java:
```diff
@@ -0,0 +1,277 @@
+        }
+    }
+
+    public static class CriteriaBuilder {
```
For consistency, can you use the builder pattern in `ContainerConfig`? Also make sure the access modifiers for classes and methods are as strict as they can be.

davidxia commented on src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java:
```diff
@@ -3267,6 +3290,199 @@ public void testOomScoreAdj() throws Exception {
+    assertThat(tasksWithId.get(0), equalTo(task));
+  }
+
+  private ServiceSpec createServiceSpec(String serviceName) {
```
nitpick: final for parameter.

davidxia commented on src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java:
```diff
@@ -3267,6 +3290,199 @@ public void testOomScoreAdj() throws Exception {
+    final ServiceSpec spec = createServiceSpec(randomName());
+    assertThat(sut.listTasks().size(), is(0));
+    final ServiceCreateResponse response = sut.createService(spec, new ServiceCreateOptions());
+    Thread.sleep(2000); // to give it a while to spin containers
```
I'm not a fan of these sleeps. We should use something like https://github.com/awaitility/awaitility, but since we have them littered throughout the tests already, we can do that in another PR.

davidxia commented:
@ketan Thanks so much for your patience. I took a closer look. I think we should be good to once you address my comments.

ketan commented on src/main/java/com/spotify/docker/client/messages/swarm/RestartPolicy.java:
```diff
@@ -0,0 +1,123 @@
+@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+public class RestartPolicy {
+
+    public static final String RESTART_POLICY_NONE = "none";
```
The docs here https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/#/create-a-service specify just these 3 policies.

davidxia commented:
replaced by https://github.com/spotify/docker-client/pull/514
